### PR TITLE
[WIP] Caching + Simplified model

### PR DESCRIPTION
**Pull request by OfirArviv**

### CODE REVIEW

File: tupa/classifiers/nn/neural_network.py [hunk 1]
```diff
@@ -77,14 +78,19 @@ def __init__(self, *args, **kwargs):
             self.bert_embedding_len = 1024 if "large" in self.config.args.bert_model else 768
 
             self.last_weights = ""
+
+            if self.config.args.cache_bert:
```
Comment (danielhers) [inline tupa/classifiers/nn/neural_network.py]: What's these attributes' value when `use_bert==True` but `cache_bert==False`?

File: tupa/classifiers/nn/neural_network.py [hunk 2]
```diff
@@ -189,47 +195,61 @@ def get_empty_values(self, key):
...
-
-        indexed_tokens = self.tokenizer.convert_tokens_to_ids(bert_tokens)
-        tokens_tensor = self.torch.tensor([indexed_tokens])
-        tokens_tensor = tokens_tensor.to('cuda')
-
-        with self.torch.no_grad():
-            encoded_layers, _ = self.bert_model(tokens_tensor)
-        assert len(encoded_layers) == self.bert_layers_count, "Invalid BERT layer count %s" % len(encoded_layers)
+        def bert_cache_token_key(full_sentence: str, layer: int):
+            return f'{full_sentence}_{layer}'
```
Comment (danielhers) [inline tupa/classifiers/nn/neural_network.py]: I'd use a tuple instead of a string

Comment (danielhers) [reply tupa/classifiers/nn/neural_network.py]: Or a nested dict to avoid the `any` below

**Comment by danielhers** [comment]
Hi @OfirArviv, is this PR still relevant?

**Comment by OfirArviv** [comment]
If I remembered correctly we wanted to use BERT directly without the birnn to improve the runtime.
Not sure its relevant, as I'm not sure if a little improvement in the runtime will affect it that much. WDYT?

**Comment by danielhers** [comment]
I think it's worth being done with - there were just a few comments I left,
no?

ons. 9. dec. 2020 19.51 skrev OfirArviv <notifications@github.com>:

**Comment by OfirArviv** [comment]
I will run it locally to verify it then. Been too long for me to 100% trust it

**Comment by danielhers** [comment]
@OfirArviv I'm going to archive this repository and continue using only https://github.com/huji-nlp/tupa to avoid duplication, so please create a PR there for these changes if you think they are worth integrating.

**Pull request closed without merging**